### PR TITLE
Don't print a newline every time output is flushed in node

### PR DESCRIPTION
**Pull request by hoodmane**
When printing to the browser console, it's not possible to print a partial line. When printing in node, it is. This fixes `flush` when in node to behave like expected of a terminal application.

Ideally, if we're in node we ought to default to preserving `isatty` for the standard streams.

### CODE REVIEW

**Comment by hoodmane** [comment]
Without this change:
```
../../test_dir/test_pyo3_pure.py::test_static 
PASSED                     [ 33%]
../../test_dir/test_pyo3_pure.py::test_class 
PASSED                      [ 66%]
../../test_dir/test_pyo3_pure.py::test_function 
PASSED                   [100%]
```

With this change:
```
../../test_dir/test_pyo3_pure.py::test_static PASSED                     [ 33%]
../../test_dir/test_pyo3_pure.py::test_class FAILED                      [ 66%]
../../test_dir/test_pyo3_pure.py::test_function PASSED                   [100%]
```

**Comment by kripken** [comment]
Is the problem that `flush` is called, the program exits, but there is still stuff unprinted? `-sEXIT_RUNTIME` should fix that, if so.

I am a little worried about changing node behavior to differ from the web. The closer we can keep those things the easier some debugging tasks get. So hopefully there is another option here.

**Comment by hoodmane** [comment]
I'm generally a bit confused by the behavior of flush in Emscripten.  Consider the following program:
```C
#include <stdio.h>
#include <unistd.h>

int main(){
    for(int i = 0; i < 10; i++){
        printf(".");
        fflush(stdout);
        sleep(1);
    }
    printf("\n");
}
```
compiled with gcc it prints a dot every second. Compiled with emcc and run with node it prints nothing and then prints the dots all at once. I want it to print a dot every second. Is there any way to opt into this behavior?

**Comment by kripken** [comment]
Right now there isn't a way to opt into that, correct. And no good way to do it on the web (or for that matter in deno and other places). This is an annoying limitation of `console.*`.

If you do absolutely need that, then overriding stdout in JS and writing to the node-specific API would probably be best, I think. Adding an example of that to the docs would be a good idea actually.

**Comment by hoodmane** [comment]
One issue is that if you set e.g., `Module['stdin']`, then `isatty` will be false. I want it to be true.

**Comment by hoodmane** [comment]
Well something like this works:

```js
function make_tty_ops(stream){
  return {
    // get_char has 3 particular return values:
    // a.) the next character represented as an integer
    // b.) undefined to signal that no data is currently available
    // c.) null to signal an EOF
    get_char(tty) {
      if (!tty.input.length) {
        var result = null;
        var BUFSIZE = 256;
        var buf = Buffer.alloc(BUFSIZE);
        var bytesRead = fs.readSync(process.stdin.fd, buf, 0, BUFSIZE, -1);
        if (bytesRead === 0) {
          return null;
        }
        result = buf.slice(0, bytesRead);
        tty.input = Array.from(result);
      }
      return tty.input.shift();
    },
    put_char(tty, val) {
      try {
        if(val !== null){
          tty.output.push(val);
        }
        if (val === null || val === 10) {
          process.stdout.write(Buffer.from(tty.output));
          tty.output = [];
        }
      } catch(e){
        console.warn(e);
      }
    },
    flush(tty) {
      if (!tty.output || tty.output.length === 0) {
        return;
      }
      stream.write(Buffer.from(tty.output));
      tty.output = [];
    }
  };
}

function setupStreams(FS, TTY){
  let mytty = FS.makedev(FS.createDevice.major++, 0);
  let myttyerr = FS.makedev(FS.createDevice.major++, 0);
  TTY.register(mytty, make_tty_ops(process.stdout))
  TTY.register(myttyerr, make_tty_ops(process.stderr))
  FS.mkdev('/dev/mytty', mytty);
  FS.mkdev('/dev/myttyerr', myttyerr);
  FS.unlink('/dev/stdin');
  FS.unlink('/dev/stdout');
  FS.unlink('/dev/stderr');
  FS.symlink('/dev/mytty', '/dev/stdin');
  FS.symlink('/dev/mytty', '/dev/stdout');
  FS.symlink('/dev/myttyerr', '/dev/stderr');
  FS.closeStream(0);
  FS.closeStream(1);
  FS.closeStream(2);
  var stdin = FS.open('/dev/stdin', 0);
  var stdout = FS.open('/dev/stdout', 1);
  var stderr = FS.open('/dev/stderr', 1);
  console.log(stdin.fd, stdout.fd, stderr.fd);
}
```

See 
https://github.com/samuelcolvin/pydantic-core/pull/129

File: tests/cstdio/test_remove.cpp [hunk 1]
```diff
@@ -57,7 +57,7 @@ void test() {
   err = std::remove("dir");
   assert(!err);
 
-  std::cout << "success";
```
Comment (sbc100) [inline tests/cstdio/test_remove.cpp:60]: Can you submit these two test changes seperately?  I think they are useful on their own.

**Comment by kripken** [comment]
> One issue is that if you set e.g., Module['stdin'], then isatty will be false. I want it to be true.

Interesting. Maybe we could add an easy way to define custom streams as `isatty`? Say if `.isatty` is set on the JS stream function, or some better way..?

**Comment by curiousdannii** [comment]
Does stdout using console.log currently say it's isatty?

**Comment by hoodmane** [comment]
Yeah, the reason is here:
https://github.com/emscripten-core/emscripten/blob/a0ed598fbc41e0c1e1958136d257d843c89a7306/src/library_fs.js#L1392
If `Module.stdin` / `stdout` / `stderr` is passed, it creates a new device with `istty` false, if it's not passed then it symlinks stdin/stdout/stderr to a tty.

**Comment by curiousdannii** [comment]
Would it make sense that a browser's isatty should be false? Whereas on Node it should use process.stdout, and return whatever the real isatty is?

If it worked like that I'd feel that @kripken's concern that Node's behaviour different from browsers would be tempered - instead it would be that in a tty everything works properly, and if not in a tty then don't rely on any particular behaviour for flush.

**Comment by hoodmane** [comment]
It's true that the browser console isn't so much like a tty. It won't interpret colors, it isn't possible to update in place, in fact it is only possible to output a whole line at a time, and there's no real way to read from it.

**Pull request closed without merging**